test/common_nn.py

```diff
@@ -2490,6 +2535,60 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
     output = output.to(dt)
     return output
 
+
+r""" input:
+        [[[0., 1., 2.],
+          [3., 4., 5.]]]
+output:
+        [[[2., 0., 1., 2., 0.],
+          [5., 3., 4., 5., 3.]]]
+"""
+
+
+def padding1d_circular(input):
+    return torch.cat([input[:, :, -1:], input,
+                      input[:, :, 0:1]], dim=2)
```

Comment:
In fact, your reference implementations seem to only implement the pad=1 case. Perhaps this was intentional? But it doesn't seem like we're getting very much utility from the reference implementation in this case.

One possible alternative is to use the implementation that looked like `torch.cat(input, input, input)` and then view out the bits we didn't need. Or we could just drop the reference implementation entirely; I agree it doesn't seem necessary in this case.

Reply:
I can add more test cases with and without reference implementation